### PR TITLE
Add UseOSCIndicator setting to enable progress indicator in terminal

### PR DESCRIPTION
**Pull request by SteveL-MSFT**
# PR Summary

Add new `$PSStyle.Progress.UseOSCIndicator` setting to use conemu's OSC indicator to show progress in the terminal window, which is also supported by Windows Terminal.  Since this is a custom OSC that isn't widely supported, it defaults to `$false`.  Since this is an advanced setting not supported by many terminals, had to give it a technical name.  iTerm2, for example, uses `ESC]9;` as a notification OSC sequence so you get a pop-up dialog instead of progress.

This is enabled as part of `PSAnsiProgress` experimental feature

image

cc @dhowett

## PR Checklist

- [x] [PR has a meaningful title](https://github.com/PowerShell/PowerShell/blob/master/.github/CONTRIBUTING.md#pull-request---submission)
    - Use the present tense and imperative mood when describing your changes
- [x] [Summarized changes](https://github.com/PowerShell/PowerShell/blob/master/.github/CONTRIBUTING.md#pull-request---submission)
- [x] [Make sure all `.h`, `.cpp`, `.cs`, `.ps1` and `.psm1` files have the correct copyright header](https://github.com/PowerShell/PowerShell/blob/master/.github/CONTRIBUTING.md#pull-request---submission)
- [x] This PR is ready to merge and is not [Work in Progress](https://github.com/PowerShell/PowerShell/blob/master/.github/CONTRIBUTING.md#pull-request---work-in-progress).
    - If the PR is work in progress, please add the prefix `WIP:` or `[ WIP ]` to the beginning of the title (the `WIP` bot will keep its status check at `Pending` while the prefix is present) and remove the prefix when the PR is ready.
- **[Breaking changes](https://github.com/PowerShell/PowerShell/blob/master/.github/CONTRIBUTING.md#making-breaking-changes)**
    - [ ] None
    - **OR**
    - [x] [Experimental feature(s) needed](https://github.com/MicrosoftDocs/PowerShell-Docs/blob/staging/reference/6/Microsoft.PowerShell.Core/About/about_Experimental_Features.md)
        - [x] Experimental feature name(s): `PSAnsiProgress`
- **User-facing changes**
    - [ ] Not Applicable
    - **OR**
    - [x] [Documentation needed](https://github.com/PowerShell/PowerShell/blob/master/.github/CONTRIBUTING.md#pull-request---submission)
        - [x] Issue filed: https://github.com/MicrosoftDocs/PowerShell-Docs/issues/7304
- **Testing - New and feature**
    - [x] N/A or can only be tested interactively
    - **OR**
    - [ ] [Make sure you've added a new test if existing tests do not effectively test the code changed](https://github.com/PowerShell/PowerShell/blob/master/.github/CONTRIBUTING.md#before-submitting)
- **Tooling**
    - [x] I have considered the user experience from a tooling perspective and don't believe tooling will be impacted.
    - **OR**
    - [ ] I have considered the user experience from a tooling perspective and opened an issue in the relevant tool repository. This may include:
        - [ ] Impact on [PowerShell Editor Services](https://github.com/PowerShell/PowerShellEditorServices) which is used in the [PowerShell extension](https://github.com/PowerShell/vscode-powershell) for VSCode
        (which runs in a different PS Host).
            - [ ] Issue filed: 
        - [ ] Impact on Completions (both in the console and in editors) - one of PowerShell's most powerful features.
            - [ ] Issue filed: 
        - [ ] Impact on [PSScriptAnalyzer](https://github.com/PowerShell/PSScriptAnalyzer) (which provides linting & formatting in the editor extensions).
            - [ ] Issue filed: 
        - [ ] Impact on [EditorSyntax](https://github.com/PowerShell/EditorSyntax) (which provides syntax highlighting with in VSCode, GitHub, and many other editors).
            - [ ] Issue filed:

### CODE REVIEW

**Comment by bergmeister** [comment]
Whilst one wouldn't want things special cased and hard-coded to terminals, one can actually detect if PowerShell is hosted within ConEmu via environment variables.
Therefore what do you think about a best effort approach of trying to detect well known terminals like ConEmu or the Windows terminal and have a different default if and only if it can be detected with confidence.

**Comment by SteveL-MSFT** [comment]
@bergmeister does ConEmu have a big enough customer base to warrant such code?  Also, does every version of ConEmu support it?  In the case of Windows Terminal, we'd have to check that it's newer than a specific version to turn it on by default.

**Comment by JustinGrote** [comment]
@SteveL-MSFT I worry about this being a slippery slope of the team having to support all kinds of terminal quirks long-term, when it's easy enough to do the detection in a user profile.

This kind of detection could also easily be done in a separate module that manages PSStyle, I don't think it should be part of the core codebase however.

**Review by daxian-dbw — APPROVED** [review]

**Comment by daxian-dbw** [comment]
@SteveL-MSFT It would be nice if you can add screenshot or GIF to show how this looks in Windows Terminal.

File: src/Microsoft.PowerShell.ConsoleHost/host/msh/ConsoleHostUserInterfaceProgress.cs [hunk 1]
```diff
@@ -92,6 +99,21 @@ class ConsoleHostUserInterface : System.Management.Automation.Host.PSHostUserInt
...
+                    int percentComplete = record.PercentComplete;
+                    if (percentComplete < 0)
+                    {
+                        // Write-Progress allows for negative percent complete, but not greater than 100
+                        // but OSC sequence is limited from 0 to 100.
+                        percentComplete = 0;
+                    }
+
+                    // OSC sequence to turn on progress indicator
+                    // https://github.com/microsoft/terminal/issues/6700
```
Comment (daxian-dbw) [inline src/Microsoft.PowerShell.ConsoleHost/host/msh/ConsoleHostUserInterfaceProgress.cs:113]: Is there an official document for the OSC sequences supported in Windows Terminal?

Comment (SteveL-MSFT) [reply src/Microsoft.PowerShell.ConsoleHost/host/msh/ConsoleHostUserInterfaceProgress.cs:113]: None I could find other than the github issue

**Comment by SteveL-MSFT** [comment]
@daxian-dbw added a gif, you can see it in the upper left of the tab.  It's just a little circle that fills in based on % complete

**Review by anmenaga — APPROVED** [review]
LGTM

**Comment by anmenaga** [comment]
Changed PR title so that changelog generation script won't have problems with it.

**Comment by sba923** [comment]
Given that the OSC progress indicator is a "write-only" thing, there are a number of scenarios where the result is not what one would expect IMVHO:

1. `a;b;c;d` one-liner where `a` sets the indicator to "indeterminate" using `OSC9;4;3;100BEL`, `b` launches a script that uses `Write-Progress`, `c` performs some time-consuming actions, and `d` resets the indicator using `OSC9;4;0;100BEL`: the "indeterminate" indicator is overwritten by the `Write-Progress` so during the execution of `c`  the indicator is not the "indeterminate" one but the "0%" one
2. If a script shows its progress using `Write-Progress` and uses `Compress-Archive`, the indicator set by the former is overwritten by the one set by the latter, so the indicator that's visible most of the time is 0% since that's what it's reset to at the end of the `Compress-Archive`

This can be experimented with something like this:

```powershell
param([int] $Count = 10, [int] $Delay = 10, [switch] $Archive)

for ($i=0; $i -lt $Count; $i++)
{
    Write-Host ("{0}/{1}" -f $i, $Count)
    Write-Progress -Activity 'Activity' -Status 'Status' -PercentComplete (100 * ($i+1) / $Count)
    if ($Archive)
    {
        Remove-Item c:/tmp/foo.zip -ea 0
        Compress-Archive -destination c:/tmp/foo.zip C:\Users\steph\OneDrive\Images\*.cr2
    }
    Start-Sleep -Seconds $Delay
}
Write-Progress -Activity 'Activity' -Status 'Status' -Completed

```

I think what we need is some global stack-based encapsulation of the indicator, where the latest desired indicator is "pushed" to the top of the stack, and the indicator is restored to the previous state on a "pop" operation.

**Comment by DHowett** [comment]
This needs to be solved for all nested and coterminous progress operations regardless of OSC indication. OSC indication being mixed up is only one symptom of the underlying issue. :smile:

**Comment by sba923** [comment]
> This needs to be solved for all nested and coterminous progress operations regardless of OSC indication. OSC indication being mixed up is only one symptom of the underlying issue. 😄

Where is discussed / tracked?

Are the `a` / `d` parts of my scenario #1 covered?

**Pull request merged**